### PR TITLE
[c2][serialized_tests] Refactor tests part 1

### PR DESCRIPTION
**Pull request by ajyu**
Followup to [the serialized test framework](https://github.com/pytorch/pytorch/pull/10594)

Round 1 for refactoring tests, starting alphabetically. I added some functionality, so I wanted to send out some of these initial changes sooner.

I'm skipping all tests that don't explicitly call assertReferenceChecks. Some tests directly call np.allclose, and others are simply TestCase (rather than HypothesisTestCase).

1. Start alphabetically producing serialized outputs for test functions, annotating those we want to include with `@serialized_test_util.given`. So far I've only added one test per operator, but this already does seem to add quite a few tests.
2. Add functionality to allow us to generate outputs using pytest by adding pytest argument options. This allows us to skip adding a `__main__` function to quite a few tests.
3. Catch any exceptions generating the gradient operator and skip serializing/reading it, since certain operators don't have gradients.
4. Add functionality to better handle jagged array inputs, which numpy doesn't handle very well. We simply explicitly do the conversion to dtype=object.
5. Make only one file per test function, rather than 4, to reduce the number of files in the github repo.

I also noticed that there is some hypothesis handling that makes `@serialized_test_util.given` not compatible with adding more hypothesis decorators on top. For example, there are tests that do 
```
@settings(...)
@given(...)
def test_my_stuff(...)
```
But there is a hypothesis handler that explicitly checks that `@given` is called below `@settings`, so we cannot refactor this to `@serialized_test_util.given`. I've just avoided decorating these kinds of tests for now, I hope that's alright.

### CODE REVIEW

**Comment by ajyu** [comment]
@houseroad @ilia-cher @BIT-silence @apaszke for review

**Review by houseroad — COMMENTED** [review]
Can we generate a table in the md doc which stats the test cases by the script?

File: caffe2/python/operator_test/adagrad_test.py [hunk 1]
```diff
@@ -15,19 +15,21 @@
...
+class TestAdagrad(serial.SerializedTestCase):
     @staticmethod
     def ref_row_wise_adagrad(param_in, mom_in, grad, lr, epsilon):
         mom_out = mom_in + np.mean(np.square(grad))
         grad_adj = lr * grad / (np.sqrt(mom_out) + epsilon)
         param_out = param_in + grad_adj
         return (param_out, mom_out)
 
-    @given(inputs=hu.tensors(n=3),
+    @serial.given(inputs=hu.tensors(n=3),
```
[the remark this answers is not in the document]
Comment (ajyu) [reply caffe2/python/operator_test/adagrad_test.py]: If there is no gradient op, then we don't serialize or compare the gradient. We still compare the output.

File: caffe2/python/serialized_test/README.md [hunk 2]
```diff
@@ -4,9 +4,15 @@ Major functionality lives in `serialized_test_util.py`
...
+2. Change the `@given` decorator to `@serialized_test_util.given`. This runs a seeded hypothesis test instance which will generate outputs if desired in addition to the unseeded hypothesis tests normally run.
+3. [Optional] Add (or change a call of `unittest.main()` to) `testWithArgs` in `__main__`. This allows you to generate outputs using `python caffe2/python/operator_test/my_test.py -G`.
+4.  Run your test `python -m pytest caffe2/python/operator_test/my_test.py -G` to generate serialized outputs. They will live in `caffe2/python/serialized_test/data/operator_test`, one npz file per test function. Use `-O` to change the output directory.
+5. Thereafter, runs of the test without the flag will load serialized outputs and gradient operators for comparison against the seeded run. The comparison is done as long as you have a call to assertReferenceChecks. If for any reason the seeded run's inputs are different (this can happen with different hypothesis versions or different setups), then we'll run the serialized inputs through the serialized operator to get a runtime output for comparison.
```
[the remark this answers is not in the document]
Comment (ajyu) [reply caffe2/python/serialized_test/README.md]: Sure, I'll add that comment.

**Review by houseroad — APPROVED** [review]

**Pull request closed without merging**